tests/functional/functional_test.go

```diff
@@ -0,0 +1,339 @@
+	Expect(err).NotTo(HaveOccurred())
+	kubeconfigData, err := clientcmd.Write(infraKubeconfig)
+	Expect(err).NotTo(HaveOccurred())
+	s.Data["kubeconfig"] = kubeconfigData
+	_, err = tenantSetup.client.CoreV1().Secrets(tenantSetup.namespace).Create(context.Background(), &s, metav1.CreateOptions{})
+	if !errors.IsAlreadyExists(err) {
+		Expect(err).NotTo(HaveOccurred())
+	}
+
+	cm := v1.ConfigMap{}
+	cmData := bytes.NewReader(generated.MustAsset("deploy/configmap.yaml"))
+	err = yaml.NewYAMLToJSONDecoder(cmData).Decode(&cm)
+	Expect(err).NotTo(HaveOccurred())
+	cm.Data["infraClusterNamespace"] = c.namespace
+	_, err = tenantSetup.client.CoreV1().ConfigMaps(tenantSetup.namespace).Create(context.Background(), &cm, metav1.CreateOptions{})
+	if !errors.IsAlreadyExists(err) {
+		Expect(err).NotTo(HaveOccurred())
+	}
+}
+
+func deployStorageClass(tenantSetup TenantCluster) {
+	sc := storagev1.StorageClass{}
+	storageClassData := bytes.NewReader(generated.MustAsset("deploy/example/storageclass.yaml"))
+	err := yaml.NewYAMLToJSONDecoder(storageClassData).Decode(&sc)
+	Expect(err).NotTo(HaveOccurred())
+	_, err = tenantSetup.client.StorageV1().StorageClasses().Create(context.Background(), &sc, metav1.CreateOptions{})
+	if !errors.IsAlreadyExists(err) {
+		Expect(err).NotTo(HaveOccurred())
+	}
+}
+
+func deployPVC(tenantSetup TenantCluster) {
+	pvc := v1.PersistentVolumeClaim{}
+	pvcData := bytes.NewReader(generated.MustAsset("deploy/example/storage-claim.yaml"))
+	err := yaml.NewYAMLToJSONDecoder(pvcData).Decode(&pvc)
+	Expect(err).NotTo(HaveOccurred())
+	_, err = tenantSetup.client.CoreV1().PersistentVolumeClaims(tenantSetup.namespace).Create(context.Background(), &pvc, metav1.CreateOptions{})
+	if !errors.IsAlreadyExists(err) {
+		Expect(err).NotTo(HaveOccurred())
+	}
+}
+
+func deployTestPod(tenantSetup TenantCluster) error {
+	testPod := v1.Pod{}
+	podData := bytes.NewReader(generated.MustAsset("deploy/example/test-pod.yaml"))
+	err := yaml.NewYAMLToJSONDecoder(podData).Decode(&testPod)
+	Expect(err).NotTo(HaveOccurred())
+	_, err = tenantSetup.client.CoreV1().Pods(tenantSetup.namespace).Create(context.Background(), &testPod, metav1.CreateOptions{})
+	if !errors.IsAlreadyExists(err) {
+		Expect(err).NotTo(HaveOccurred())
+	}
+
+	testPodUpTimeout, cancelFunc := context.WithTimeout(context.Background(), time.Minute*20)
+	defer cancelFunc()
+	_, err = clientwatch.UntilWithSync(
+		testPodUpTimeout,
+		cache.NewListWatchFromClient(tenantSetup.client.CoreV1().RESTClient(), "pods",
+			tenantSetup.namespace, fields.OneTermEqualSelector("metadata.name", testPod.Name)),
+		&v1.Pod{},
+		nil,
+		func(event apiwatch.Event) (bool, error) {
+			switch event.Type {
+			case apiwatch.Added, apiwatch.Modified:
+				p, ok := event.Object.(*v1.Pod)
+				if !ok {
+					Fail("Failed fetching the test pod")
+				}
+				if p.Status.Phase == v1.PodRunning {
+					return true, nil
+				}
+			default:
+				return false, nil
+			}
+
+			return false, nil
+		},
+	)
```

Comment:
I think that this part should be a separate function.

Reply:
you didn't say which part exactly but I guess the until with sync?

Reply:
Yes